utils/http_client.go

```diff
@@ -19,12 +19,18 @@ type (
 )
 
 // NewHubHTTPClient is creating the default client
-func NewHubHTTPClient() HTTPClient {
+func NewHubHTTPClient() HubHTTPClient {
 	return HubHTTPClient{
 		httpClient: &http.Client{},
 	}
 }
 
+// WithTransport replaces the underlying HTTP client transport
+func (hc HubHTTPClient) WithTransport(transport http.RoundTripper) HTTPClient {
```

Comment:
You can just implement your own implementation of the `HTTPClient` with this functionality so I don't see the big gain here.

Reply:
No worries, I just wanted to avoid duplicating the `utils.handleResponse` in my code.

Reply:
Look at #14 and see if that would solve your problem

Reply:
Thanks. It's a good start but it doesn't work for me because I also measure request time and this is only a pre-request hook.
It would either need to be a pre/post pair (with context passed between each) or a proper chained middleware (each hook responsible for calling the next).

FWIW, I use a generic logging HTTP transport that looks something like:

```go
var client = &http.Client{Transport: &loggingRoundTripper{}}

type loggingRoundTripper struct {
	http.RoundTripper
}

func (t loggingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
	start := time.Now()
	res, err := t.RoundTrip(req)
	log := zerolog.Ctx(req.Context()).With().
		Str("method", req.Method).
		Str("path", req.URL.Path).
		Dur("timer", time.Since(start)).
		Logger()
	if err != nil {
		log.Error().Err(err).Msg("Client request failed")
	} else {
		log.Debug().Str("status", http.StatusText(res.StatusCode)).Msg("Client request completed")
	}
	return res, err
}
```

Reply:
The simplest solution if this is all a bit overkill would be to just embed the inner `http.Client` in `utils.HubHTTPClient` so that it's accessible to anyone who doesn't mind a cast.

```go
type HubHTTPClient struct {
	*http.Client
}

h := NewNotificationHub("...", "...")
c := utils.NewHubHTTPClient().(*utils.HubHTTPClient)
c.Transport = &myTransport{}
h.SetHTTPClient(c)
```

I'd be totally fine with that.

Reply:
Ok I see. The way I solved the same problem is that I measure the whole request to the API wrapper since it is basically no logic. Just marshalling, an http request and then marshalling again. So the timing and duration of the whole request is just nano seconds longer than the http request which is by far the slowest part of the process.

The reason why I hesitate to add this feature is that I imagine this module as a black box around the API and that you shouldn't have to care to much about the actual HTTP implementation itself. We do also support a custom httpClient already.

But if you insist, Im happy to consider a middeware kind of implementation since I think that is the most flexible and keeps my mental image intact.

Reply:
Not insisting, this PR was just a suggestion that would allow me to easily add transparent metrics similar to other client libraries I'm using (`go-pg`, `go-redis`, `resty`, `grpc`, etc). I'm fine with using a custom hub http client as you've suggested 👍.

What are your thoughts on embedding the `http.Client` in `utils.HubHTTPClient` rather than an unexported field? I'll update this PR for your consideration.

Reply:
If I'm not misstaken, embedding the http client will require a refactor of the test suite which doesn't sound like fun :)

Reply:
No, the test suite only mocks the `utils.HTTPClient` interface. The `utils.HubHTTPClient` struct is not used in tests, otherwise Travis would have failed when I pushed this change an hour ago.